src/coreComponents/dataRepository/wrapperHelpers.hpp

```diff
@@ -98,7 +99,47 @@ real64 const * getPointerToComponent( R1Tensor const & var, int const component
 
 } // namespace internal
 
+template< typename T >
+class ArrayDimLabels
+{
+public:
+
+  void set( integer const, Span< string const > )
+  {
+    GEOSX_ERROR( "Dimension labels are only available in Array wrappers" );
+  }
+
+  Span< string const > get( integer const ) const
+  {
+    GEOSX_ERROR( "Dimension labels are only available in Array wrappers" );
+    return {};
+  }
+};
 
+template< typename T, int NDIM, typename PERM >
+class ArrayDimLabels< Array< T, NDIM, PERM > >
+{
+public:
+
+  void set( integer const dim, Span< string const > labels )
+  {
+    GEOSX_ERROR_IF_LT( dim, 0 );
+    GEOSX_ERROR_IF_GE( dim, NDIM );
+    m_values[dim].resize( labels.size() );
+    std::copy( labels.begin(), labels.end(), m_values[dim].begin() );
+  }
```

Comment:
Maybe a check to make sure that the the size of the `dim` dimension is equal to the number of labels?

I know this is a chicken or egg situation. You are setting the labels prior to the resize of the actual array, but then the resize should check to make sure it has the appropriate resize for a dimension corresponding to the number of labels in `m_dimLabels.m_values`.  Maybe this is already done and I missed it.

Reply:
Yeah, that's exactly the doubt I've had. Since resizing the array does _not_ happen through `Wrapper` but instead through a direct reference, there is no way to always enforce the invariant. So I decided to check "eventual consistency" only at usage, i.e. in output code. 

Adding the check here can prevent _some_ potential bugs, but it will also prevent the following usage (actually seen in code):
```c++
subRegion.getWrapper< array2d< real64, compflow::LAYOUT_COMP > >( viewKeyStruct::globalCompDensityString() ).
      setDimLabels( 1, fluid.componentNames() ).
      reference().resizeDimension< 1 >( m_numComponents );
```
Here I cannot reorder the calls because once you get `reference()`, wrapper method chaining is lost. I'll have to unchain it and store reference to wrapper in a local variable, or just move setting labels to a different function entirely.